### ISSUE TITLE
`@overload` and `TypeVar` interact badly

### ISSUE THREAD

**Issue by colehaus** [microsoft/pyright#3889]
I would expect all of the following code to type check. However, the body of `a` does not type check. Surprisingly, leaving off the type declaration on the left-hand side (as in `b`) allows the function to type check even thought the type declaration in `a` exactly matches the type inferred in `b`.

```py
from typing import TypeVar, overload

A = TypeVar("A")

def foo(a: A) -> A:
    return a

@overload
def bar(b: str) -> str: ...
@overload
def bar(b: int) -> int: ...

def a(x: int | str) -> None:
    _: int | str = bar(foo(x)) # Error here

def b(x: int | str) -> None:
    _ = bar(foo(x))

def c(x: int | str) -> None:
    _: int | str = bar(x)

def d(x: int | str) -> None:
    _: int | str = foo(x)
```

 The error raised by the pyright CLI in version 1.1.269 is
```
 "error: Argument of type "int | str" cannot be assigned to parameter "a" of type "A@foo" in function "foo"                  
    Type "int | str" cannot be assigned to type "str"
      Type "int | str" cannot be assigned to type "str"
        "int" is incompatible with "str" (reportGeneralTypeIssues)
```

The code does type check in mypy 0.971.

**Comment by erictraut** [comment]
Thanks for the bug report. This will be fixed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.270, which I just published. It will also be included in a future release of pylance.
